### DISCUSSION TITLE
how to use esbuild --bundle with nbb bundle output

### DISCUSSION THREAD

**Discussion by dukefong8** [babashka/nbb#260] [Q&A]
**Is your feature request related to a problem? Please describe.**
nbb bundle use a top level 'await loadString(...)' in output .mjs file. esbuild currently does'nt support top leve await. I tried  command `npx esbuild --bundle ink_demo.mjs --outfile=ink_demo1.mjs --platform=node --target=node16 --format=esm` , but running `node ink_demo1.mjs` gave error like

```
file:///home/hacker/dev/scripts/ink_demo1.mjs:12
  throw new Error('Dynamic require of "' + x + '" is not supported');
        ^

Error: Dynamic require of "assert" is not supported
    at file:///home/hacker/dev/scripts/ink_demo1.mjs:12:9
    at node_modules/signal-exit/index.js (file:///home/hacker/dev/scripts/ink_demo1.mjs:7623:17)
    at __require2 (file:///home/hacker/dev/scripts/ink_demo1.mjs:15:52)
    at node_modules/restore-cursor/index.js (file:///home/hacker/dev/scripts/ink_demo1.mjs:7773:22)
    at __require2 (file:///home/hacker/dev/scripts/ink_demo1.mjs:15:52)
    at node_modules/cli-cursor/index.js (file:///home/hacker/dev/scripts/ink_demo1.mjs:7786:25)
    at __require2 (file:///home/hacker/dev/scripts/ink_demo1.mjs:15:52)
    at node_modules/ink/build/log-update.js (file:///home/hacker/dev/scripts/ink_demo1.mjs:7825:40)
    at __require2 (file:///home/hacker/dev/scripts/ink_demo1.mjs:15:52)
    at node_modules/ink/build/ink.js (file:///home/hacker/dev/scripts/ink_demo1.mjs:75279:40)

Node.js v18.7.0
```

**Describe the solution you'd like**
It'd be nice if nbb bundle has an optional not to use top level await.
 
**Describe alternatives you've considered**
I tried manually remove await in front of loadString from nbb bundle's output ink_demo.mjs. Then use command `npx esbuild --bundle ink_demo.mjs --outfile=ink_demo1.js --platform=node --target=node16` (note that esbuild output uses cjs module this time).  `node ink_demo1.js` does work.

**Additional context**

**Comment by borkdude** [comment]
Hello @FongHou,

You can find the documentation on bundling here:

https://github.com/babashka/nbb/tree/main/doc/bundle

**Reply by borkdude** [reply]
See:


```
npx esbuild out.mjs --bundle --minify --platform=node --outfile=dist/index.mjs --format=esm
```

The `format=esm` argument supports top level await.

**Reply by dukefong8** [reply]
Thanks @borkdude!

I did have format=esm as you can see in my test above. However, at least the esbuild ---format=esm node bundle doesn't work for ink_demo.cljs. Not sure if it's an esbuild issue or not. 

On the other hand, I found out that esbuild --format=cjs node bundle does work if nbb bundle does NOT put await in front of loadString(...) line, as you can see in below screenshot. So my question is  about whether that top level await is really needed or not.

![image](https://user-images.githubusercontent.com/13973254/193431053-52ab9ba2-5e2e-42b5-9826-ee1503495ea2.png)

**Reply by borkdude** [reply]
> I did have format=esm as you can see in my test above.

Sorry I missed that.
Do you have a project I can clone locally so I can try this out?
